draft-ietf-quic-transport.md

```diff
@@ -2277,11 +2278,21 @@ linked by any other entity.
 At any time, endpoints MAY change the Destination Connection ID they send to a
 value that has not been used on another path.
 
-An endpoint MUST use a new connection ID if it initiates connection migration as
-described in {{initiating-migration}} or probes a new network path as described
-in {{probing}}.  An endpoint MUST use a new connection ID in response to a
-change in the address of a peer if the packet with the new peer address uses an
-active connection ID that has not been previously used by the peer.
+An endpoint MUST NOT reuse a connection ID when sending from more than one local
+address, for example when initiating connection migration as described in
+{{initiating-migration}} or when probing a new network path as described in
+{{probing}}.
+
+Similarly, an endpoint MUST NOT reuse a connection ID when sending to more than
+one destination address.  However, if an endpoint receives packets from a new
+source address with the same destination connection ID, it MAY continue to use
+the current connection ID with the new address.
```

Comment:
This text is really confusing. We have a direct prohibition followed by an unexplained exception. Can we provide some reasoning here?

Reply:
How about:

> However, if a peer experiences NAT rebinding, an endpoint could receive packets from its peer using a new source address with the same destination connection ID.  If a peer is first to use a connection ID on a new network path and that connection ID was originally used on another network path, an endpoint MAY also use a connection ID that it used on the original network path.

That is much more wordy, but it might be clearer.  The point is to say that you don't reuse a connection ID unless your peer reuses first.

There is a wrinkle here in the case that an endpoint has multiple valid addresses (think preferred address).  In that case, we don't want to allow reuse as that implies a bunch of other things.  The fix there might be to add:

> This exception that allows the use of a connection ID on multiple network paths only applies when the remote address changes and the local address remains constant.

Reply:
Do we actually have to make this exception? I haven't been tracking all of the conversations, but why not just have endpoints use a new CID in this case.

In general an endpoint MUST limit the use of a connection ID to a single pair of remote and local addresses. This prevents linkage between network paths. However, if an endpoint may experiences unexpected NAT rebinding, that endpoint's peer may receive a packet from a new remote address but reusing an existing connection ID, thus apparently violating this rule linking those paths. [TODO: Something about how you have to accept this?]
As an exception to the rule above, if an endpoint sees a given connection ID reused from a new remote address, it MAY also reuse its corresponding connection ID when sending to that address, as long as it sends from the same local address.

Reply:
I like the framing of the last sentence you have there, although it's interesting to think through the question about if we really need an exception here. 

In theory, the problem with changing for this is that a NAT rebinding can cause us to "burn" a CID, making it so that instead of having 2 linked CIDs, we now have 3 linked CIDs. As long as we retire the old one (which is already in the text) once we change to the new one, we should be able to keep a steady supply of those around, although it would mean that someone in the middle of the network could fiddle with things to keep us generating and exchanging more and more CIDs. 

So, the idea is that if someone on the network changes without your knowledge, we've already linked those two 5-tuples by the fact that you've got the same CID in one direction, there's no harm in continuing to keep using that CID in the other direction. 

It does seem better to keep doing that, rather than opening ourselves up to letting someone on the network arbitrarily churn through CIDs.

This would mean the text becomes much like what you have in the TODO above, essentially we're saying that the _sender_ of the packet MUST follow these restrictions, but if you see that your peer did not, then \<the text you have above\>.

Reply:
I am comfortable with either outcome here.  @erickinnear, do you want to try to integrate these comments in something approximating the shape you describe?

Reply:
Yup! Working on an update for these this afternoon.

Reply:
Updated to fold this in, also note that much of this explanation (and further requirements on behavior in the event of a NAT rebinding, such as potentially needing to validate the new path) is present in the immediately following paragraph: 
```
These requirements regarding connection ID reuse apply only to the sending of
packets, as unintentional changes in path without a change in connection ID are
possible.  For example, after a period of network inactivity, NAT rebinding
might cause packets to be sent on a new path when the client resumes sending.
An endpoint responds to such an event as described in {{migration-response}}.
```
Updated in f7a9c3b

Reply:
Thanks for doing that Eric.  It looks good to me.